---
output: html_document
editor_options: 
  chunk_output_type: console
---
Load Libraries
```{r, echo = FALSE, collapse = TRUE, message=FALSE}
library("ggplot2")
library("DESeq2")
library('dplyr')
library('stringr')
library('rtracklayer')
library('msigdbr')
library('clusterProfiler')
library('gridExtra')
```
 
```{r}
#make function to plot informative MA plots called ggplotMA: 
  ## data: results from DESeq objects
  ## genes: if user wishes to label top x # most significantly changing genes, can specify number 1:x
  ## Title: string that specifies title of MA Plot
ggplotMA <- function(data, genes, Title){
  data$Color <- ifelse(data$padj > 0.05, "not significant", ifelse(abs(data$log2FoldChange) > 1, "padj < 0.05 & LFC > 1", "padj < 0.05"))
MAdefault <- function(data, Title){
ggplot(as.data.frame(data), aes(x=baseMean, y=log2FoldChange, color=Color))+
scale_colour_manual(values=c("grey", "pink", "red"))+
geom_point(data=as.data.frame(data), aes(x=baseMean, y=log2FoldChange), size=0.2)+ 
scale_x_continuous(trans='log10')+
ggtitle(paste(Title, "MAPlot"))+
xlab("log10 Base Mean")+
theme_classic(base_size = 20)+
ylim(c(min(-max(data$log2FoldChange),min(data$log2FoldChange)), max(max(data$log2FoldChange),-min(data$log2FoldChange))))+ #centers y axis around 0
labs(color="Significance")
  }
  
MAdefault(data, Title) 
}
``` 
 
```{r}
genecounts <- read.table(file="~/linker-histone/RNA-seq/rep2/hg38-ERCC-UMI-alignment/all-genefeaturecounts.txt", header = TRUE)
colnames(genecounts) <- c("GeneID","Chr","Start","End","Strand","Length","low1","low2","low3","scr1","scr2","scr3")
counts <- dplyr::select(genecounts,low1,low2,low3,scr1,scr2,scr3)
rownames(counts) <- gsub('\\.[0-9]*$', '',genecounts$GeneID)

colData <- as.data.frame(x = c("low", "low", "low", "scr", "scr", "scr"))
colnames(colData) <- c("genotype")
colData$genotype <- as.factor(colData$genotype)
colData$genotype <- factor(colData$genotype,levels = c("scr", "low"))
rownames(colData) <- colnames(counts)

dds <- DESeqDataSetFromMatrix(countData = counts, colData = colData, design = ~genotype)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

ctrlGenes <- grep("GERCC", rownames(dds))
dds <- estimateSizeFactors(dds, controlGenes=ctrlGenes)

de.dds <- DESeq(dds)
write.csv(assay(de.dds), "/lustre/fs4/risc_lab/scratch/iduba/linker-histone/RNA-seq/rep2/hg38-ERCC-UMI-alignment/DESeq2_results/dds_raw_counts_matrix.csv")
plotDispEsts(de.dds)

resLFC <- lfcShrink(de.dds, coef="genotype_low_vs_scr", type="normal")
resLFC.noNA <- na.omit(resLFC)
ggplotMA(resLFC.noNA, 1:20, "WTvslow (LFC)")

resLFC.df <- as.data.frame(resLFC.noNA)
resLFC.df$type <- ifelse(resLFC.df$padj < 0.05 & resLFC.df$log2FoldChange > 1, "Up", ifelse(resLFC.df$padj < 0.05 & resLFC.df$log2FoldChange < -1, "Down", "Not significant"))
write.csv(resLFC.df, "/lustre/fs4/risc_lab/scratch/iduba/linker-histone/RNA-seq/rep2/hg38-ERCC-UMI-alignment/DESeq2_results/WTvslow.csv")

WTvslowup <- subset(subset(resLFC, padj < 0.05), log2FoldChange > 1)
WTvslowdown <- subset(subset(resLFC, padj < 0.05), log2FoldChange < -1)
WTvslownochange <- subset(subset(resLFC, log2FoldChange < 1), log2FoldChange > -1)
write.csv(WTvslowup, "/lustre/fs4/risc_lab/scratch/iduba/linker-histone/RNA-seq/rep2/hg38-ERCC-UMI-alignment/DESeq2_results/WTvslowup.csv")
write.csv(WTvslowdown, "/lustre/fs4/risc_lab/scratch/iduba/linker-histone/RNA-seq/rep2/hg38-ERCC-UMI-alignment/DESeq2_results/WTvslowdown.csv")
write.csv(WTvslownochange, "/lustre/fs4/risc_lab/scratch/iduba/linker-histone/RNA-seq/rep2/hg38-ERCC-UMI-alignment/DESeq2_results/WTvslownochange.csv")

lowBMup <- subset(WTvslowup, baseMean < median(WTvslowup$baseMean))
highBMup <- subset(WTvslowup, baseMean >= median(WTvslowup$baseMean))
write.csv(lowBMup, "/lustre/fs4/risc_lab/scratch/iduba/linker-histone/RNA-seq/rep2/hg38-ERCC-UMI-alignment/DESeq2_results/WTvslowup-lowbasemean.csv")
write.csv(highBMup, "/lustre/fs4/risc_lab/scratch/iduba/linker-histone/RNA-seq/rep2/hg38-ERCC-UMI-alignment/DESeq2_results/WTvslowup-highbasemean.csv")
```

## GO enrichment analysis with msigdb GO:BP database:
```{r}
Ens_genes_v107 <- import(con="/lustre/fs4/risc_lab/store/jyeung/Homo_sapiens.GRCh38.107.gff3") 
Ens_genes_v107 <- Ens_genes_v107 %>% as.data.frame() %>% dplyr::filter(!is.na(gene_id)) %>% dplyr::distinct(Name, description, gene_id) 
colnames(Ens_genes_v107) <- c("gene_id","description", "ensembl_id") 

GRCh38.p14 <- import(con="/lustre/fs4/risc_lab/store/jyeung/ncbi-genomes-2022-09-11/GCF_000001405.40_GRCh38.p14_genomic.gtf") 

GRCh38.p14$entrezid <- gsub('.*\\GeneID:','', GRCh38.p14$db_xref) 
GRCh38.p14 <- GRCh38.p14[, c(1:16, 31)] 
GRCh38.p14_genesdf <- GRCh38.p14 %>% as.data.frame() %>% dplyr::distinct(gene_id, description, entrezid) 
GRCh38.p14_genesdf <- GRCh38.p14_genesdf[-(grep("HGNC|MIM|RFAM|miRBase", GRCh38.p14_genesdf$entrezid)), ] 

genes_anno_type <- merge(Ens_genes_v107, GRCh38.p14_genesdf[, c(1,3)], by="gene_id", all=T)

de.dds_rlog <- rlog(de.dds)
universe <- genes_anno_type[match(row.names(de.dds_rlog), genes_anno_type$ensembl_id), ]

  ## gene= vector of genes to test for GO enrichment (gene ensembl format), 
  ## universe=background gene set
  ## Database=database containing groups of genes to GO annotation term e.g. msigdb GO:BP

getGOenrichment <-function(gene,universe,Database){
  gene_entrezID <- genes_anno_type[match(gene, genes_anno_type$ensembl_id), ]
  gene_go <- enricher(gene_entrezID$entrezid, TERM2GENE=Database, universe=universe) #enrichment analysis with hypergeometric test
 # gene_go <- setReadable(gene_go, OrgDb = org.Hs.eg.db, keyType="ENTREZID") #convert enrichment results from ENTREZ ID format to SYMBOL for easier interpretation
return(gene_go)
} 

# msigdb GO Biological Processes database
msigdbr_df_GO_BP <- as.data.frame(msigdbr(species = "Homo sapiens", category="C5", subcategory ="GO:BP"))
msigdbr_t2g_GO_BP = msigdbr_df_GO_BP %>% dplyr::distinct(gs_name, entrez_gene) %>% as.data.frame()
msigdbr_t2g_GO_BP$gs_name <- sub("GOBP_", "", msigdbr_t2g_GO_BP$gs_name) 
msigdbr_t2g_GO_BP$gs_name <- gsub("_", " ", msigdbr_t2g_GO_BP$gs_name)

genes_in_sig <- list()
GO_enrichment <- list()
dotplots <- list()
sig_dfs <- list(WTvslowup, WTvslowdown)
names_genes <- list("upregulated","downregulated")
for(i in 1:2){
 genes_in_sig[[i]] <- rownames(sig_dfs[[i]])  
 names(genes_in_sig)[i] <- names_genes[i]
 GO_enrichment[[i]] <- getGOenrichment(genes_in_sig[[i]], as.character(universe$entrezid),msigdbr_t2g_GO_BP) # do GO enrichment analysis with background gene set set as all gene expressed, database used is msigdb GO:BP
 dotplots[[i]] <- dotplot(GO_enrichment[[i]], showCategory=10, font.size=12, title=paste(names_genes[i], "genes enriched GO terms")) # make dotplot of top 50 most significantly enriched GO terms 
}
grid.arrange(grobs=dotplots) # plot dotplots of enriched GO terms of each cluster all at once 

yus_1_up <- read.table(file="/lustre/fs4/home/iduba/linker-histone/RNA-seq/genes-of-interest/yus-rep1-common-up.st-genebody-cscore.bedgraph", header=TRUE)
GO_yus_1_up <- getGOenrichment(yus_1_up$gene_ID,as.character(universe$entrezid),msigdbr_t2g_GO_BP)
dotplot(GO_yus_1_up, showCategory=10, font.size=12, title="Yusufova-rep1 common up genes")

yus_2_up <- read.table(file="/lustre/fs4/home/iduba/linker-histone/RNA-seq/genes-of-interest/yus-rep2-common-up.st-genebody-cscore.bedgraph", header=TRUE)
GO_yus_2_up <- getGOenrichment(yus_2_up$gene_ID,as.character(universe$entrezid),msigdbr_t2g_GO_BP)
dotplot(GO_yus_2_up, showCategory=10, font.size=12, title="Yusufova-rep2 common up genes")

wil_1_up <- read.table(file="/lustre/fs4/home/iduba/linker-histone/RNA-seq/genes-of-interest/wil-rep1-common-up.st-genebody-cscore.bedgraph", header=TRUE)
GO_wil_1_up <- getGOenrichment(wil_1_up$gene_ID,as.character(universe$entrezid),msigdbr_t2g_GO_BP)
dotplot(GO_wil_1_up, showCategory=10, font.size=12, title="Willcockson-rep1 common up genes")

wil_2_up <- read.table(file="/lustre/fs4/home/iduba/linker-histone/RNA-seq/genes-of-interest/wil-rep2-common-up.st-genebody-cscore.bedgraph", header=TRUE)
GO_wil_2_up <- getGOenrichment(wil_2_up$gene_ID,as.character(universe$entrezid),msigdbr_t2g_GO_BP)
dotplot(GO_wil_2_up, showCategory=10, font.size=12, title="Yusufova-rep2 common up genes")


rep1_2_up <- read.table(file="/lustre/fs4/home/iduba/linker-histone/RNA-seq/genes-of-interest/rep1-2-common-up.st-genebody-cscore.bedgraph", header=TRUE)
GO_1_2_up <- getGOenrichment(rep1_2_up$gene_ID,as.character(universe$entrezid),msigdbr_t2g_GO_BP)
dotplot(GO_1_2_up, showCategory=10, font.size=12, title="K562-reps1-2 common up genes enriched GO terms")
```
